components/nvs/include/nvs.hpp

```diff
@@ -0,0 +1,116 @@
+            // NVS partition was truncated and needs to be erased
+            // Retry nvs_flash_init
+            ESP_ERROR_CHECK(nvs_flash_erase());
+            err = nvs_flash_init();
+        }
+        if (err != ESP_OK) {
+            logger_.error("NVS INIT FAILED");
+            ec = std::make_error_code(std::errc::protocol_error);
+        }
+    }
+
+    template<typename T> void set_var(const char *ns_name, const char *key, T value, std::error_code &ec) {
+        init(ec);
+        if (ec)
+            return;
+        esp_err_t err;
+        std::unique_ptr<nvs::NVSHandle> handle = nvs::open_nvs_handle(ns_name, NVS_READWRITE, &err);
+        if (err != ESP_OK) {
+            logger_.error("Error {%s} opening NVS handle!", esp_err_to_name(err));
+            ec = std::make_error_code(std::errc::protocol_error);
+            return;
+        }
+        err = handle->set_item(key, value);
+        if (err != ESP_OK) {
+            ec = std::make_error_code(std::errc::protocol_error);
+            logger_.error("Error {%s} writing to NVS!", esp_err_to_name(err));
+            return;
+        }
+        err = handle->commit();
+        if (err != ESP_OK) {
+            ec = std::make_error_code(std::errc::protocol_error);
+            logger_.error("Error {%s} committing to NVS!", esp_err_to_name(err));
+        }
+        return;
+    }
+
+    template<typename T> void get_var(const char *ns_name, const char *key, T &value, T default_value, std::error_code &ec) {
+        value = default_value;
+        init(ec);
+        if (ec)
+            return;
+        esp_err_t err;
+        std::unique_ptr<nvs::NVSHandle> handle = nvs::open_nvs_handle(ns_name, NVS_READWRITE, &err);
+        if (err != ESP_OK) {
+            logger_.error("Error {%s} opening NVS handle!", esp_err_to_name(err));
+            ec = std::make_error_code(std::errc::protocol_error);
+            return;
+        }    
+        T readvalue;
+        err = handle->get_item(key, readvalue);
+        switch (err) {
+        case ESP_OK:
+            value = readvalue;
+            break;
+        case ESP_ERR_NVS_NOT_FOUND:
+            logger_.warn("The value is not initialized yet! Set to: {}", default_value);
+            err = handle->set_item(key, default_value);
+            if (err != ESP_OK) {
+                logger_.error("Error {%s} writing to NVS!", esp_err_to_name(err));
+                ec = std::make_error_code(std::errc::protocol_error);
+                return;
+            }
+            err = handle->commit();
+            if (err != ESP_OK) {
+                logger_.error("Error {%s} committing to NVS!", esp_err_to_name(err));
+                ec = std::make_error_code(std::errc::protocol_error);
+            }
+            break;
+        default:
+            logger_.error("Error {%s} reading!", esp_err_to_name(err));
+            ec = std::make_error_code(std::errc::protocol_error);
+        }
+        return;
+    }
```

Comment:
I might break this up into two functions: `get_var(...)` which doesn't take a default value and does not do a `set_item(...)`/`commit()`, and a `get_or_set(...)` (or some other name that makes sense) which does what you've implemented here. Currently a user of the API has no way to see if there is a key in that namespace without creating it. Similarly, might be good to have a simple `exists(...)` function which returns true or false and uses the `find_key(...)` API.